src/txn/txn_rollback_to_stable.c

```diff
@@ -980,6 +980,109 @@ __rollback_to_stable_btree(WT_SESSION_IMPL *session, wt_timestamp_t rollback_tim
     return (ret);
 }
 
+/*
+ * __rollback_evict_exclusive_apply --
+ *     Apply an eviction helper to all open dhandles.
+ */
+static int
+__rollback_evict_exclusive_apply(WT_SESSION_IMPL *session, bool *files_acquired,
+  size_t files_acquired_len, int (*dhandle_func)(WT_SESSION_IMPL *, bool *, size_t))
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DATA_HANDLE *dhandle;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    for (dhandle = NULL;;) {
+        WT_WITH_HANDLE_LIST_WRITE_LOCK(session, WT_DHANDLE_NEXT(session, dhandle, &conn->dhqh, q));
+        if (dhandle == NULL)
+            break;
+
+        if (!WT_PREFIX_MATCH(dhandle->name, "file:"))
+            continue;
+
+        /* Evictions of the metadata or history store files won't cause history store operations. */
+        if (WT_IS_METADATA(dhandle) || WT_IS_HS((WT_BTREE *)dhandle->handle))
+            continue;
+
+        WT_WITH_DHANDLE(
+          session, dhandle, ret = (*dhandle_func)(session, files_acquired, files_acquired_len));
+        WT_RET(ret);
+    }
+
+    return (0);
+}
+
+/*
+ * __rollback_evict_file_exclusive_on_wrapper --
+ *     A wrapper around the "exclusive on" helper that marks the file as acquired.
+ */
+static int
+__rollback_evict_file_exclusive_on_wrapper(
+  WT_SESSION_IMPL *session, bool *files_acquired, size_t files_acquired_len)
+{
+    WT_BTREE *btree;
+
+    btree = S2BT(session);
+
+    if (btree->id >= files_acquired_len || files_acquired[btree->id])
```

Comment:
I suspect that using btree-apply would mean you could skip this. You are probably seeing multiple entries because the some of the data handles correspond to non-btree types that wrap btrees..

Reply:
This is not called on the same id twice. The reason this particular check is here is that I believe we should be able to call `__rollback_evict_exclusive_on/off` repeatedly without breaking anything. If we've already acquired everything, trying to do it again should just be a no-op.

More generally, this mechanism exists because if we run into an error halfway through acquiring, we'll only want to release the ones that we acquired. There's no way to tell what we've acquired just by inspecting the btree so we need to keep that state somewhere which is what I'm doing with `files_acquired`.